File: implementations/php/lib/Eval/PawnStructure.php
<?php

namespace Chess\Eval;

use Chess\Board;

final class PawnStructure {
    public static function evaluate(Board $board): int {
        $whitePawns = self::collectPawns($board, CHESS_WHITE);
        $blackPawns = self::collectPawns($board, CHESS_BLACK);

        return self::evaluateColor($board, CHESS_WHITE, $whitePawns, $blackPawns)
            - self::evaluateColor($board, CHESS_BLACK, $blackPawns, $whitePawns);
    }

    private static function evaluateColor(Board $board, int $color, array $pawns, array $enemyPawns): int {
        $score = 0;
        $fileCounts = array_fill(0, 8, 0);
        foreach ($pawns as [$row, $col]) {
            $fileCounts[$col]++;
        }

        foreach ($pawns as [$row, $col]) {
            if ($fileCounts[$col] > 1) {
                $score -= 20;
            }

            $hasAdjacentPawn = ($col > 0 && $fileCounts[$col - 1] > 0) || ($col < 7 && $fileCounts[$col + 1] > 0);
            if (!$hasAdjacentPawn) {
                $score -= 15;
            }

            if (self::isPassedPawn($color, $row, $col, $enemyPawns)) {
                $score += Tables::passedPawnBonusByRank(self::rankFromWhitePerspective($color, $row));
            }

            if (self::isPawnChain($board, $color, $row, $col)) {
                $score += 10;
            }

            if (self::isConnectedPawn($board, $color, $row, $col)) {
                $score += 5;
            }
        }

        return $score;
    }

    private static function collectPawns(Board $board, int $color): array {
        $pawns = [];
        for ($row = 0; $row < 8; $row++) {
            for ($col = 0; $col < 8; $col++) {
                [$piece, $pieceColor] = $board->get_piece($row, $col);
                if ($piece === CHESS_PAWN && $pieceColor === $color) {
                    $pawns[] = [$row, $col];
                }
            }
        }
        return $pawns;
    }

    private static function isPassedPawn(int $color, int $row, int $col, array $enemyPawns): bool {
        foreach ($enemyPawns as [$enemyRow, $enemyCol]) {
            if (abs($enemyCol - $col) > 1) {
                continue;
            }
            if ($color === CHESS_WHITE && $enemyRow < $row) {
                return false;
            }
            if ($color === CHESS_BLACK && $enemyRow > $row) {
                return false;
            }
        }

        return true;
    }

    private static function isPawnChain(Board $board, int $color, int $row, int $col): bool {
        $supportRow = $color === CHESS_WHITE ? $row + 1 : $row - 1;
        if ($supportRow < 0 || $supportRow >= 8) {
            return false;
        }

        foreach ([-1, 1] as $deltaCol) {
            $supportCol = $col + $deltaCol;
            if ($supportCol < 0 || $supportCol >= 8) {
                continue;
            }
            [$piece, $pieceColor] = $board->get_piece($supportRow, $supportCol);
            if ($piece === CHESS_PAWN && $pieceColor === $color) {
                return true;
            }
        }

        return false;
    }

    private static function isConnectedPawn(Board $board, int $color, int $row, int $col): bool {
        foreach ([-1, 1] as $deltaCol) {
            $adjacentCol = $col + $deltaCol;
            if ($adjacentCol < 0 || $adjacentCol >= 8) {
                continue;
            }
            [$piece, $pieceColor] = $board->get_piece($row, $adjacentCol);
            if ($piece === CHESS_PAWN && $pieceColor === $color) {
                return true;
            }
        }

        return false;
    }

    private static function rankFromWhitePerspective(int $color, int $row): int {
        return $color === CHESS_WHITE ? 7 - $row : $row;
    }
}

File: implementations/haskell/src/Eval/PawnStructure.hs
module Eval.PawnStructure where

import Types
import Board
import Eval.Tables
import Data.List (group, sort)

evaluatePawnStructure :: GameState -> Int
evaluatePawnStructure gs =
  scoreFor White - scoreFor Black
  where
    scoreFor color =
      doubledPenalty color + isolatedPenalty color + passedBonus color + chainBonus color + connectedBonus color
    pawns color = [sq | sq <- allSquares, getPiece gs sq == Just (Piece color Pawn)]
    allSquares = [Square col row | col <- [0..7], row <- [0..7]]

    doubledPenalty color =
      let files = sort [c | Square c _ <- pawns color]
          excess = sum [max 0 (length grp - 1) | grp <- group files]
      in excess * (-20)

    isolatedPenalty color =
      let ps = pawns color
      in (-15) * length [sq | sq@(Square c _) <- ps, not (hasPawnOnAdjacentFile ps c)]

    passedBonus color =
      sum [passedPawnValue color sq | sq <- pawns color, isPassedPawn color sq]

    chainBonus color =
      10 * length [sq | sq <- pawns color, isPawnDefendedByPawn color sq]

    connectedBonus color =
      5 * length [sq | sq <- pawns color, isConnectedPawn color sq]

    hasPawnOnAdjacentFile ps file = any (\(Square c _) -> abs (c - file) == 1) ps

    isPassedPawn color (Square col row) =
      let enemy = opponentColor color
          rows = if color == White then [row + 1 .. 7] else [0 .. row - 1]
          files = [max 0 (col - 1) .. min 7 (col + 1)]
      in not (any (\(c, r) -> getPiece gs (Square c r) == Just (Piece enemy Pawn)) [(c, r) | c <- files, r <- rows])

    passedPawnValue White (Square _ row) = clampedLookup passedPawnBonusByRank row
    passedPawnValue Black (Square _ row) = clampedLookup passedPawnBonusByRank (7 - row)

    isPawnDefendedByPawn White (Square col row) = any (hasPawnColor White) [(col - 1, row - 1), (col + 1, row - 1)]
    isPawnDefendedByPawn Black (Square col row) = any (hasPawnColor Black) [(col - 1, row + 1), (col + 1, row + 1)]

    isConnectedPawn color (Square col row) =
      any hasPawn [(col - 1, row), (col + 1, row)]
      where
        hasPawn (c, r)
          | c < 0 || c > 7 || r < 0 || r > 7 = False
          | otherwise = getPiece gs (Square c r) == Just (Piece color Pawn)

    hasPawnColor color (c, r)
      | c < 0 || c > 7 || r < 0 || r > 7 = False
      | otherwise = getPiece gs (Square c r) == Just (Piece color Pawn)
